$($TIME_DATE) is never changing. $($TIME_CLOCK) totally wrong,

date never changes, i guess its just using the date wehn the script was started ? date will never change on new uploads

![jFY34ma](https://github.com/jenslys/autovod/assets/157306043/9329234d-bf5c-4b8a-9ad7-8feb7d3c94fe)


also the time_clock seems totally wrong, its grabbing the time from twitch ? i guess booth should use the local/host time/date

so here is my default config.

#!/bin/bash
# shellcheck disable=SC2034
#! Do not edit this file directly, instead create a new file with the name of the streamer you want to record. (e.g: asmongold.config)
set -a

#* Source
STREAM_SOURCE="twitch" #* Options: twitch, kick, youtube

#* Upload provider
UPLOAD_SERVICE="youtube" #* Options: youtube, rclone, restream, local

#* Twitch metadata settings:
API_CALLS="true" #? Enable this to fetch additional metadata from the Twitch API. API_URL must be set to a valid API URL.
API_URL="https://tw-test.vercel.app/"        #? API for fetching stream metadata. (e.g: https://example.com/info/)

#? Variables that can be used in VIDEO_TITLE, YOUTUBE_VIDEO_DESCRIPTION, YOUTUBE_VIDEO_PLAYLIST, RCLONE_FILENAME and LOCAL_FILENAME.
STREAMER_TITLE="initial_title" #? The title of the stream. (This is only available if API_CALLS is set to true.) [! Do not change these values]
STREAMER_GAME="inital_game"    #? The current game the user is playing. (This is only available if API_CALLS is set to true.) [! Do not change these values]
#? $STREAMER_NAME - The name of the streamer.
#? $($TIME_DATE) - The current date (day/month/year).
#? $($TIME_CLOCK) - The current time (hour:minute:second).

#* Streamlink settings (Valid for all upload providers)
STREAMLINK_QUALITY="best"                                               #* Options: worst, 360p, 480p, 720p, 720p60, 1080p60, best
STREAMLINK_FLAGS=("--twitch-disable-reruns" "--twitch-disable-hosting" "--twitch-api-header=Authorization=OAuth ok3zgbqv9hcu3utzsi54uzqmrm4ghj") #? Options to pass to streamlink. (NB! Array not a string) (https://streamlink.github.io/cli.html#twitch)
STREAMLINK_LOGS="error"                                                 #* Options: none, error, warning, info, debug, trace, all

#* Universal Upload settings (Valid for all upload providers)
VIDEO_DURATION="12:00:00"       #? XX:XX:XX (Notice: YouTube has a upload limit of 12 hours per video).
SPLIT_INTO_PARTS="false"        #? If you want to split the video into parts, set this to true. (if this is enabled VIDEO_DURATION is ignored).
SPLIT_VIDEO_DURATION="06:00:00" #? Set the duration of each part. (XX:XX:XX)
SAVE_ON_FAIL="false"            #? If you want to save the video that failed to upload, set this to true.

#* YouTube upload settings (Only valid if UPLOAD_SERVICE is set to youtube)
VIDEO_TITLE="$STREAMER_NAME - $($TIME_DATE)"                          #? Title of the video.
VIDEO_DESCRIPTION="Uploaded using AutoVOD" #? YouTube video description.
VIDEO_VISIBILITY="unlisted"                                           #* Options: unlisted, private, public
VIDEO_PLAYLIST="$STREAMER_NAME VODs"                                  #? Playlist to upload to.

#* Rclone upload settings (Only valid if UPLOAD_SERVICE is set to rclone)
RCLONE_REMOTE="remote-name"                      #? Rclone remote to upload to.
RCLONE_DIR=""                                    #? Directory to upload to.
RCLONE_FILENAME="$STREAMER_NAME"_"$($TIME_DATE)" #? Filename of the video on the remote. (Dont use spaces in the filename, use dashes or undersores instead.)
RCLONE_FILEEXT="mkv"                             #? File extension of the video on the remote.

#* Re-Stream settings (Only valid if UPLOAD_SERVICE is set to restream)
RTMPS_URL="rtmp://a.rtmp.youtube.com/live2/"
RTMPS_STREAM_KEY=""
AUDIO_BITRATE="44100"
AUDIO_CODEC="aac"
FILE_FORMAT="flv"

#* Local settings (Only valid if UPLOAD_SERVICE is set to local)
LOCAL_FILENAME="$STREAMER_NAME"_"$($TIME_DATE)" #? Filename/filepath of the local recording. (Dont use spaces in the filename, use dashes or undersores instead.)
LOCAL_EXTENSION="mkv"                           #? File extension of the local recording.

#* Re-Encoding settings (Currently only works if UPLOAD_SERVICE is set to rclone or local)
#? Re-encodes the stream to a desired codec and quality using ffmpeg. This can be useful if you want to save space on your remote. (Only recommended if you have a high-end server.)
RE_ENCODE="false"         #? Set this to true if you want to re-encode the video.
RE_ENCODE_CODEC="libx265" #* Options: libx265, libvpx-vp9, libaom-av1 (https://ffmpeg.org/ffmpeg-codecs.html)
RE_ENCODE_CRF="25"        #* Options: 0-51 (lower is better quality, but larger file size) (https://slhck.info/video/2017/02/24/crf-guide.html)
RE_ECODE_PRESET="medium"  #* Options: ultrafast, superfast, veryfast, faster, fast, medium, slow, slower, veryslow (https://trac.ffmpeg.org/wiki/Encode/H.265)
RE_ENCODE_LOG="error"     #* Options: none, error, warning, info, debug, trace, all

set +a

--- a/AutoVOD.sh
+++ b/AutoVOD.sh
@@ -82,9 +82,15 @@ determineSource
 while true; do
 	# Store the orignal values
 	variables=("VIDEO_TITLE" "VIDEO_PLAYLIST" "VIDEO_DESCRIPTION" "RCLONE_FILENAME" "RCLONE_DIR" "LOCAL_FILENAME")
+	TIME_DATE_CACHE=$($TIME_DATE) # We don't need to execute same command repeatedly
+	TIME_CLOCK_CACHE=$($TIME_CLOCK)
 	for var in "${variables[@]}"; do
 		original_var=original_$var
 		eval "$original_var=\$$var"
+		eval "replace_var=\$$var"
+		replace_var=${replace_var//TIME_DATE/$TIME_DATE_CACHE}
+		replace_var=${replace_var//TIME_CLOCK/$TIME_CLOCK_CACHE}
+		eval "$var=\$replace_var"
 	done
 
 	fetchMetadata() {
@@ -256,26 +262,40 @@ while true; do
 		;;
 
 	"local")
-		if [ "$RE_ENCODE" == "true" ]; then
-			echo "$($CC) Re-encoding stream"
-			if ! streamlink $STREAM_SOURCE_URL $STREAMLINK_OPTIONS "${STREAMLINK_FLAGS[@]}" --stdout | ffmpeg -i pipe:0 -c:v $RE_ENCODE_CODEC -crf $RE_ENCODE_CRF -preset $RE_ECODE_PRESET -hide_banner -loglevel $RE_ENCODE_LOG -f matroska $LOCAL_FILENAME >/dev/null 2>&1; then
-				echo "$($CC) ffmpeg failed re-encoding the stream"
-			else # If the stream was re-encoded
-				echo "$($CC) Stream re-encoded as $LOCAL_FILENAME"
+		TEMP_FILE=$(mktemp stream.XXXXXX)
+
+		if ! streamlink $STREAM_SOURCE_URL $STREAMLINK_OPTIONS "${STREAMLINK_FLAGS[@]}" -o - >$TEMP_FILE; then
+			echo "$($CC) streamlink failed saving the stream"
+			if [ "$SAVE_ON_FAIL" == "true" ] && [ "$(stat -c %b $TEMP_FILE)" != "0" ]; then
+				#? Save the temp file if rclone fails
+				NEW_LOCAL_FILENAME="${LOCAL_FILENAME}_failed.$LOCAL_EXTENSION"
+				mv "$TEMP_FILE" "$NEW_LOCAL_FILENAME" # Rename the local file
+				echo "$($CC) Local failed file renamed to $NEW_LOCAL_FILENAME"
 			fi
+			rm -f $TEMP_FILE
 		else
-			# If you just want to save the stream locally to your machine
-			if ! streamlink $STREAM_SOURCE_URL $STREAMLINK_OPTIONS "${STREAMLINK_FLAGS[@]}" -o - >"$LOCAL_FILENAME.$LOCAL_EXTENSION"; then
-				echo "$($CC) streamlink failed saving the stream to disk"
-				if [ "$SAVE_ON_FAIL" == "true" ]; then
-					#? Save the temp file if rclone fails
-					NEW_LOCAL_FILENAME="${LOCAL_FILENAME}_failed.$LOCAL_EXTENSION"
-					mv "$LOCAL_FILENAME.$LOCAL_EXTENSION" "$NEW_LOCAL_FILENAME" # Rename the local file
-					echo "$($CC) Local failed file renamed to $NEW_LOCAL_FILENAME"
+			if [ "$RE_ENCODE" == "true" ]; then
+				echo "$($CC) Re-encoding stream"
+				if ! ffmpeg -i "$TEMP_FILE" -c:v $RE_ENCODE_CODEC -crf $RE_ENCODE_CRF -preset $RE_ECODE_PRESET -hide_banner -loglevel $RE_ENCODE_LOG -f matroska $LOCAL_FILENAME >/dev/null 2>&1; then
+					echo "$($CC) ffmpeg failed re-encoding the stream"
+				else # If the stream was re-encoded
+					echo "$($CC) Stream re-encoded as $LOCAL_FILENAME"
 				fi
 			else
-				echo "$($CC) Stream saved to disk as $LOCAL_FILENAME.$LOCAL_EXTENSION"
-				TIME_DATE_CHECK=$($TIME_DATE)
+				# If you just want to save the stream locally to your machine
+				if ! mv "$TEMP_FILE" "$LOCAL_FILENAME.$LOCAL_EXTENSION"; then
+					echo "$($CC) failed to move temp file to disk." # May be disk R/O mode?
+					if [ "$SAVE_ON_FAIL" == "true" ]; then
+						# Failed to move. so, we cannot do anything :(
+						echo "$($CC) Local failed file renamed to $TEMP_FILE"
+					else
+						rm -f $TEMP_FILE
+					fi
+				else
+					echo "$($CC) Stream saved to disk as $LOCAL_FILENAME.$LOCAL_EXTENSION"
+					rm -f $TEMP_FILE
+					TIME_DATE_CHECK=$($TIME_DATE)
+				fi
 			fi
 		fi
 		;;

--- a/default.config
+++ b/default.config
@@ -17,8 +17,8 @@ API_URL=""        #? API for fetching stream metadata. (e.g: https://example.com
 STREAMER_TITLE="initial_title" #? The title of the stream. (This is only available if API_CALLS is set to true.) [! Do not change these values]
 STREAMER_GAME="inital_game"    #? The current game the user is playing. (This is only available if API_CALLS is set to true.) [! Do not change these values]
 #? $STREAMER_NAME - The name of the streamer.
-#? $($TIME_DATE) - The current date (day/month/year).
-#? $($TIME_CLOCK) - The current time (hour:minute:second).
+#? TIME_DATE - The current date (day/month/year).
+#? TIME_CLOCK - The current time (hour:minute:second).
 
 #* Streamlink settings (Valid for all upload providers)
 STREAMLINK_QUALITY="best"                                               #* Options: worst, 360p, 480p, 720p, 720p60, 1080p60, best
@@ -29,10 +29,10 @@ STREAMLINK_LOGS="error"                                                 #* Optio
 VIDEO_DURATION="12:00:00"       #? XX:XX:XX (Notice: YouTube has a upload limit of 12 hours per video).
 SPLIT_INTO_PARTS="false"        #? If you want to split the video into parts, set this to true. (if this is enabled VIDEO_DURATION is ignored).
 SPLIT_VIDEO_DURATION="06:00:00" #? Set the duration of each part. (XX:XX:XX)
-SAVE_ON_FAIL="false"            #? If you want to save the video that failed to upload, set this to true.
+SAVE_ON_FAIL="false"            #? If you want to save the non-empty video that failed to upload, set this to true.
 
 #* YouTube upload settings (Only valid if UPLOAD_SERVICE is set to youtube)
-VIDEO_TITLE="$STREAMER_NAME - $($TIME_DATE)"                          #? Title of the video.
+VIDEO_TITLE="$STREAMER_NAME - TIME_DATE"                          #? Title of the video.
 VIDEO_DESCRIPTION="Uploaded using https://github.com/jenslys/AutoVOD" #? YouTube video description.
 VIDEO_VISIBILITY="unlisted"                                           #* Options: unlisted, private, public
 VIDEO_PLAYLIST="$STREAMER_NAME VODs"                                  #? Playlist to upload to.
@@ -40,7 +40,7 @@ VIDEO_PLAYLIST="$STREAMER_NAME VODs"                                  #? Playlis
 #* Rclone upload settings (Only valid if UPLOAD_SERVICE is set to rclone)
 RCLONE_REMOTE="remote-name"                      #? Rclone remote to upload to.
 RCLONE_DIR=""                                    #? Directory to upload to.
-RCLONE_FILENAME="$STREAMER_NAME"_"$($TIME_DATE)" #? Filename of the video on the remote. (Dont use spaces in the filename, use dashes or undersores instead.)
+RCLONE_FILENAME="$STREAMER_NAME"_"TIME_DATE" #? Filename of the video on the remote. (Dont use spaces in the filename, use dashes or undersores instead.)
 RCLONE_FILEEXT="mkv"                             #? File extension of the video on the remote.
 
 #* Re-Stream settings (Only valid if UPLOAD_SERVICE is set to restream)
@@ -51,7 +51,7 @@ AUDIO_CODEC="aac"
 FILE_FORMAT="flv"
 
 #* Local settings (Only valid if UPLOAD_SERVICE is set to local)
-LOCAL_FILENAME="$STREAMER_NAME"_"$($TIME_DATE)" #? Filename/filepath of the local recording. (Dont use spaces in the filename, use dashes or undersores instead.)
+LOCAL_FILENAME="$STREAMER_NAME"_"TIME_DATE" #? Filename/filepath of the local recording. (Dont use spaces in the filename, use dashes or undersores instead.)
 LOCAL_EXTENSION="mkv"                           #? File extension of the local recording.
 
 #* Re-Encoding settings (Currently only works if UPLOAD_SERVICE is set to rclone or local)

